### PR TITLE
Improve formula generators

### PR DESCRIPTION
**Pull request by matthias-wende-frequenz**
We use the recently introduced DFS in the component graph to update the pv power and the producer power formule in the logical meter.

### CODE REVIEW

File: tests/timeseries/test_logical_meter.py [hunk 1]
```diff
@@ -240,6 +240,17 @@ async def test_pv_power_no_meter(self, mocker: MockerFixture) -> None:
...
+    async def test_pv_power_no_pv_components(self, mocker: MockerFixture) -> None:
+        """Test the pv power formula without having any pv components."""
+        mockgrid = MockMicrogrid(grid_meter=True)
+        await mockgrid.start(mocker)
+
+        logical_meter = microgrid.logical_meter()
+        pv_power_receiver = logical_meter.pv_power.new_receiver()
+
+        await mockgrid.mock_resampler.send_non_existing_component_value()
+        assert (await pv_power_receiver.receive()).value == Power.from_watts(0.0)
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/test_logical_meter.py]: You might want to compare it against power zero constructor `Power.zero()`

File: tests/timeseries/mock_microgrid.py [hunk 2]
```diff
@@ -277,8 +278,11 @@ def add_chps(self, count: int) -> None:
             )
 
             self._start_meter_streaming(meter_id)
-            self._connections.add(Connection(self._connect_to, meter_id))
-            self._connections.add(Connection(meter_id, chp_id))
+            if not no_meters:
+                self._connections.add(Connection(self._connect_to, meter_id))
+                self._connections.add(Connection(meter_id, chp_id))
+            else:
+                self._connections.add(Connection(self._connect_to, chp_id))
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/mock_microgrid.py]:
```suggestion
            if no_meters:
                self._connections.add(Connection(self._connect_to, chp_id))
            else:
                self._connections.add(Connection(self._connect_to, meter_id))
                self._connections.add(Connection(meter_id, chp_id))
```

File: RELEASE_NOTES.md [hunk 3]
```diff
@@ -22,3 +22,5 @@
 - Fix `consumer_power()` not working certain configurations.
   In microgrids without consumers and no main meter, the formula
   would never return any values.
+- Fix `pv_power` not working in setups with 2 grid meters by using DFS
```
Comment (daniel-zullo-frequenz) [inline RELEASE_NOTES.md]: I'd suggest to change `by using DFS` to something around `by using a new reliable function to search for components in the components graph`

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py [hunk 4]
```diff
@@ -34,32 +34,26 @@ def generate(self) -> FormulaEngine[Power]:
...
-            # if in the future we support additional producers, we need to add them here
-            if component_graph.is_chp_chain(successor) or component_graph.is_pv_chain(
-                successor
-            ):
-                if not first_iteration:
-                    builder.push_oper("+")
+        for component in producer_components:
+            if not first_iteration:
+                builder.push_oper("+")
+            first_iteration = False
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py]: I'd suggest to replace this flag by using the same technique than in pv power formula, by enumerating the components and checking idx greater than 0 to push the operator "+". This is just for consistency

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py [hunk 5]
```diff
@@ -34,36 +36,37 @@ def generate(self) -> FormulaEngine[Power]:
...
+        )
 
-                first_iteration = False
+        if not producer_components:
+            _logger.warning(
+                "Unable to find any producer components in the component graph. "
+                "Subscribing to the resampling actor with a non-existing "
+                "component id, so that `0` values are sent from the formula."
+            )
+            # If there are no PV inverters, we have to send 0 values at the same
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py]: maybe `producer components` instead of `PV inverters`?

Comment (matthias-wende-frequenz) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py]: actually that's not true in the pv formula too. It should be pv components instead. will change that as well.

File: src/frequenz/sdk/microgrid/_graph.py [hunk 6]
```diff
@@ -266,6 +266,29 @@ def is_chp_chain(self, component: Component) -> bool:
...
+    def dfs(
+        self,
+        current_node: Component,
+        visited: Set[Component],
+        condition: Callable[[Component], bool],
+    ) -> Set[Component]:
+        """
+        Search for components that fulfill the condition in the Graph.
+
+        DFS is used for searching the graph. The graph travarsal is stopped
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/microgrid/_graph.py]: s/travarsal/traversal

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py (hunk 5 above)
Comment (matthias-wende-frequenz) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py]: done

File: src/frequenz/sdk/microgrid/_graph.py [hunk 7]
```diff
@@ -266,6 +266,29 @@ def is_chp_chain(self, component: Component) -> bool:
...
+        once a component fulfills the condition.
+
+        Args:
+            current_node: The current node to search from.
+            visited: The set of visited nodes.
+            condition: The condition function to check for.
+
+        Returns:
+            A set of component ids where the coresponding components fulfill
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/microgrid/_graph.py]: s/coresponding/corresponding

File: src/frequenz/sdk/microgrid/_graph.py (hunk 6 above)
Comment (matthias-wende-frequenz) [reply src/frequenz/sdk/microgrid/_graph.py]: ups that's wrong and is already merged

File: src/frequenz/sdk/microgrid/_graph.py (hunk 7 above)
Comment (matthias-wende-frequenz) [reply src/frequenz/sdk/microgrid/_graph.py]: I'll add an extra commit for fixing the typos!

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py (hunk 5 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_producer_power_formula.py]: right, makes sense. Thanks!

**Review by daniel-zullo-frequenz — APPROVED** [review]

**Review by daniel-zullo-frequenz — APPROVED** [review]

**Pull request merged**